### PR TITLE
feat(designer): precise scanned-STL outlines, 5mm clearance cap, free import rotation

### PR DESCRIPTION
**Pull request by andymai**
Addresses user feedback on the STL imprint feature (#2618).

## Squares instead of precise outlines (scanned STLs)

Self-scanned meshes project with dozens of speck rings (scan-noise islands) and dense, noisy contours. The old outline extraction shared one point budget and one epsilon across every ring, doubling epsilon up to 6.4mm until the total fit — so the specks ate the budget and the *main* silhouette collapsed to a near-square. Since every imported cutout carries a default clearance, the degraded outline's clearance prism then carved that square into the printed bin.

Now:
- Speck rings are filtered by area first: a ring survives if it clears 0.2% of the largest ring's area or 4mm² absolute, so a real 3mm prong on a 300mm tool is never dropped.
- The point budget (raised 2000 → 4000, server mirror updated) is split across surviving rings by perimeter, largest first.
- Per-ring epsilon soft-caps at 0.5mm — overflow drops the smallest rings whole instead of crushing the main outline.

Existing imported assets have their outlines baked in; re-uploading the STL picks up the fix.

## Clearance cap 2mm → 5mm

The cap was a hardcoded UI prop; scans need more slack than machined parts. Now `MAX_CUTOUT_CLEARANCE = 5` (all clearance shapes).

## Free rotation at import

The orientation dialog's quarter-turn-only flips become per-axis degree rotation: 15° stepper buttons, typed values, and the existing quarter-turn buttons. Rotation still composes after the deterministic auto lay-flat, so re-runs stay stable. (Rotation about the vertical axis was already free in the 2D editor via the rotate handle — unchanged.)

## Verification

- New real-WASM fixture test: a detailed round tool + 3mm prong + 60 sub-mm noise islands → specks filtered, prong kept, main ring ≥40 points, total under cap.
- Free-angle (45°) and modulo-360 rotation tests through the real pipeline.
- Dialog/hook tests for the new rotation controls; typecheck, lint, i18n checks, and the CutoutsSection/bridge/handlers/shared-generation suites (901 tests) all pass.

---
## Summary by cubic
Improves STL import outlines for scanned meshes, adds free per-axis rotation at import, and raises the cutout clearance cap to 5 mm for better fit.

- **Bug Fixes**
  - Preserve scanned-silhouette detail: filter speck rings by area; split a 4000-point cap by perimeter; soft-cap simplification at 0.5 mm and clamp epsilon ladders to the 0.5/5 mm caps; if over budget, drop the smallest rings instead of flattening the main outline.
  - Reliability and consistency: add real-WASM tests for budget-overflow ring-drop and relative speck thresholds; normalize rotation angles modulo 360; align bridge/worker APIs and tests to degrees.
  - Performance: compute ring perimeters once in outline extraction.

- **New Features**
  - Increase max insertion clearance from 2 mm to 5 mm across all clearance shapes.
  - Import dialog now supports per-axis rotation in degrees (15° stepper, typed values, plus quarter-turn buttons), composed after auto lay-flat for stable re-runs.

<sup>Written for commit 612c392cdaaf1c9fb1ba3307758c8252596fc65f. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/andymai/gridfinity-layout-tool/pull/2628?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/features/generation/worker/generators/meshImport.ts [hunk 1]
```diff
@@ -80,32 +94,74 @@ function signedArea(ring: ReadonlyArray<readonly [number, number]>): number {
...
+      });
+
+      const total = (): number => outlines.reduce((sum, ring) => sum + ring.length, 0);
+      while (outlines.length > 1 && total() > MAX_MESH_OUTLINE_POINTS) outlines.pop();
+      let epsilon = OUTLINE_EPSILON_SOFT_MAX_MM;
+      while (total() > MAX_MESH_OUTLINE_POINTS && epsilon < OUTLINE_EPSILON_HARD_MAX_MM) {
         epsilon *= 2;
-        outlines = rings.map((ring) => simplifyRdp(ring, epsilon));
+        outlines = [simplifyRdp(kept[0], epsilon)];
       }
```
[the remark this answers is not in the document]
Comment (andymai) [reply src/features/generation/worker/generators/meshImport.ts:167]: Fixed in 69316ba58 — both epsilon ladders now clamp to their caps before simplifying.

File: src/features/generation/worker/generators/meshImport.ts [hunk 2]
```diff
@@ -227,13 +283,15 @@ export async function importMeshFromStl(
...
+      Number.isFinite(deg) ? ((deg % 360) + 360) % 360 : 0;
+    const userRotation: readonly [number, number, number] = [
+      normalizeDeg(rotation.x),
+      normalizeDeg(rotation.y),
+      normalizeDeg(rotation.z),
     ];
     const laid = solid.rotate([...layFlat]);
-    const flipped = laid.rotate([...flipRotation]);
+    const flipped = laid.rotate([...userRotation]);
     laid.delete();
```
[the remark this answers is not in the document]
Comment (andymai) [reply src/features/generation/worker/generators/meshImport.ts:299]: Renamed to `rotated` in 078d5f184.

File: src/features/generation/worker/generators/meshImport.ts [hunk 3]
```diff
@@ -80,32 +94,77 @@ function signedArea(ring: ReadonlyArray<readonly [number, number]>): number {
...
+      const kept = outer
+        .filter(({ area }) => area >= speckThreshold)
+        .sort((a, b) => b.area - a.area)
+        .map(({ ring }) => ring.map(([x, y]): MeshOutlinePoint => ({ x, y })));
+
+      const totalPerimeter = kept.reduce((sum, ring) => sum + ringPerimeter(ring), 0);
+      let outlines = kept.map((ring) => {
+        const share = totalPerimeter > 0 ? ringPerimeter(ring) / totalPerimeter : 0;
+        const budget = Math.max(MIN_RING_BUDGET, Math.floor(MAX_MESH_OUTLINE_POINTS * share));
+        return simplifyRingToBudget(ring, budget);
```
[the remark this answers is not in the document]
Comment (andymai) [reply src/features/generation/worker/generators/meshImport.ts]: Fixed in 612c392cd — perimeters are computed once into an array and reused for the shares.
